tests/surrogate_test.py

```diff
@@ -0,0 +1,206 @@
+
+
+def test_surrogate_0():
+    """Test if surrogate generates exact results for test fucntion 0."""
+    surrogate_object = Surrogate([(-1, 1), (-1, 1), (-1, 1)],
+                                 SmolyakGridGenerator(
+                                     ChebyshevFirstKind.make_nested_set(2)))
+    surrogate_object.train(function_0, 'lu')
+    test_point = (0.649, 0, -0.9)
+    test_point_ = (-0.885, 1, 0.275)
+    assert numpy.allclose(surrogate_object(test_point),
+                          function_0([0.649, 0, -0.9]))
+    assert numpy.allclose(surrogate_object(test_point_),
+                          function_0([-0.885, 1, 0.275]))
+
+
+def test_surrogate_1():
+    """Test if surrogate generates exact results for test for Chebyshevs."""
+    surrogate_object = Surrogate([(-1, 1), (-1, 1)], SmolyakGridGenerator(
+        ChebyshevFirstKind.make_nested_set(2)))
+    surrogate_object.train(function_1, 'lstsq')
+    test_point = (0.649, -0.9)
+    test_point_ = (-0.885, 1)
+    assert numpy.allclose(surrogate_object(test_point),
+                          function_1([0.649, -0.9]))
+    assert numpy.allclose(surrogate_object(test_point_),
+                          function_1([-0.885, 1]))
+
+
+def test_surrogate_0_shifted():
+    """Test if surrogate generates exact results for a shifted function 0."""
+    surrogate_object = Surrogate([(-1, 1), (-1, 1), (-1, 1)],
+                                 SmolyakGridGenerator(
+                                     ChebyshevFirstKind.make_nested_set(2)))
+    surrogate_object_ = Surrogate([(0, 2), (-2, 0), (-1, 1)],
+                                  SmolyakGridGenerator(
+                                      ChebyshevFirstKind.make_nested_set(2)))
+    surrogate_object.train(function_0, 'lu')
+    surrogate_object_.train(function_0_shifted, 'lu')
+    assert numpy.allclose(surrogate_object.coefficients,
+                          surrogate_object_.coefficients)
+
+
+def test_surrogate_2():
+    """Test if surrogate generates exact results for simple 1D function."""
+    domain = (-10, 10)
+    surrogate_object = Surrogate(domain, SmolyakGridGenerator(
+        ChebyshevFirstKind.make_nested_set(2)))
+    surrogate_object.train(function_2, 'inv')
+    test_point = 0.367
+    assert numpy.allclose(surrogate_object(test_point), function_2(test_point))
+
+
+def test_train_from_data():
+    """Test if train_from_data_works."""
+    surrogate_object = Surrogate([(-5, 10), (0, 15)], SmolyakGridGenerator(
+        ChebyshevFirstKind.make_nested_set(4)))
+    surrogate_object_ = Surrogate([(-5, 10), (0, 15)], SmolyakGridGenerator(
+        ChebyshevFirstKind.make_nested_set(4)))
+    surrogate_object.train_from_data(branin_at_grids(), 'lu')
+    assert numpy.allclose(surrogate_object((8, 0.75)),
+                          branin([8, 0.75]), atol=1e-10)
+    with pytest.raises(IndexError):
+        surrogate_object_.train_from_data(branin_at_grids()[:5], 'lu')
+
```

Comment:
Should you add tests for the different solver methods? You could also parametrize the tests on this, so all will run with all solvers?